core/trino-main/src/main/java/io/trino/sql/planner/ExpressionInterpreter.java

```diff
@@ -604,14 +604,13 @@ protected Object visitInPredicate(InPredicate node, Object context)
             }
 
             if (!(value instanceof Expression)) {
-                Set<?> set = inListCache.get(valueList);
-
                 // We use the presence of the node in the map to indicate that we've already done
                 // the analysis below. If the value is null, it means that we can't apply the HashSet
                 // optimization
                 if (!inListCache.containsKey(valueList)) {
+                    Set<?> set = null;
                     if (valueList.getValues().stream().allMatch(Literal.class::isInstance) &&
-                            valueList.getValues().stream().noneMatch(NullLiteral.class::isInstance)) {
+                            valueList.getValues().stream().noneMatch(ExpressionInterpreter::expressionIsNull)) {
```

Comment:
The changes here are so that

- `RewriteVarcharConstant` doesn't choke on a `Constant` that is a null value. For this, `RewriteVarcharConstant` should be changed instead
- `FilterStatsCalculator` doesn't fail when it sees a comparison with `CAST(NULL AS double)`. This is a problem here: https://github.com/trinodb/trino/blob/67d2df9ac15f8169a492dd8e9ccc44fb879424f5/core/trino-main/src/main/java/io/trino/cost/FilterStatsCalculator.java#L413 as the code doesn't take NULL value into account. That code should be fixed instead.

Other than that, i don't understand what is the benefit of the changes in this class, so please either explain to me or revert changes in ExpressionInterpreter.

Reply:
Fixed and dropped controversial commit